src/math/ec_scalar.cpp

```diff
@@ -18,36 +18,82 @@
  */
 #include <bitcoin/system/math/ec_scalar.hpp>
 
+#include <algorithm>
 #include <bitcoin/system/math/hash.hpp>
 #include <bitcoin/system/math/elliptic_curve.hpp>
+#include <bitcoin/system/utility/serializer.hpp>
 
 namespace libbitcoin {
 namespace system {
 
-// TODO: implement equality and inequality for ec_scalar and ec_secret.
 const ec_scalar ec_scalar::zero = null_hash;
 
-// Null is valid but uninitialized object is invalid.
 ec_scalar::ec_scalar()
-  : valid_(false), secret_(null_hash)
 {
+    // Avoid initializing secret by default.
+    // Helps when you pre-allocate large arrays of values.
+}
+
+ec_scalar::ec_scalar(uint64_t value)
+{
+    *this = value;
 }
 
 ec_scalar::ec_scalar(const ec_secret& secret)
-  : valid_(true), secret_(secret)
 {
+    *this = secret;
+}
+
+ec_scalar::ec_scalar(const ec_scalar& scalar)
+{
+    *this = scalar;
+}
+
+ec_scalar::ec_scalar(ec_scalar&& scalar)
+{
+    *this = std::move(scalar);
 }
 
 // Operators.
 // ----------------------------------------------------------------------------
 
+ec_scalar& ec_scalar::operator=(uint64_t value)
+{
+    secret_.reset(new ec_secret);
```

Comment:
Unfortunately `std::make_unique` is not C++11 only 14 or later, otherwise I prefer to use that. See https://en.cppreference.com/w/cpp/memory/unique_ptr/make_unique

Reply:
Ah, thanks. We don't use unique pointer anywhere. It always seems preferable to use shared pointer and explicitly dump the pointer where desired. As it's strictly a class internal this is always possible and adds no more code. I would generally avoid its use in any public interface as well.